File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/V1/ColumnOrganizedStachExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FactSet.Protobuf.Stach.Extensions.Models;

namespace FactSet.Protobuf.Stach.Extensions.V1
{
    public class ColumnOrganizedStachExtension : IStachExtension
    {
        private readonly Package pkg;
        
        public ColumnOrganizedStachExtension(Package package)
        {
            this.pkg = package;
        }

        public List<Models.Table> ConvertToTable()
        {
            var tables = new List<Models.Table>();
            foreach (var primaryTableId in pkg.PrimaryTableIds)
            {
                tables.Add(GenerateTable(pkg, primaryTableId));
            }

            return tables;
        }

        /// <summary>
        /// The purpose of this function is to generate Table for a given table id in the provided
        /// stach data through the package.
        /// </summary>
        /// <param name="package"></param>
        /// <param name="primaryTableId"></param>
        /// <returns>Returns the generated Table from the package provided.</returns>
        private static Models.Table GenerateTable(Package package, string primaryTableId)
        {
            var primaryTable = package.Tables[primaryTableId];
            var headerId = primaryTable.Definition.HeaderTableId;
            var headerTable = package.Tables[headerId];
            var columnIds = primaryTable.Definition.Columns.Select(c => c.Id).ToList();
            var rowCount = primaryTable.Data.Rows.Count;

            var table = new Models.Table
            {
                Rows = new List<Row>(),
                Metadata = new Dictionary<string, string>()
            };

            var headerDataRowList = headerTable.Data.Rows.Select(x => x.Id).ToList();
            foreach (var headerTableSeriesDefinition in headerTable.Definition.Columns)
            {
                var headerRow = new Row {Cells = new List<string>()};
                foreach (var primaryTableSeriesDefinition in primaryTable.Definition.Columns)
                {
                    if (primaryTableSeriesDefinition.IsDimension)
                    {
                        var description = string.IsNullOrEmpty(primaryTableSeriesDefinition.Description)
                            ? primaryTableSeriesDefinition.Name
                            : primaryTableSeriesDefinition.Description;

                        headerRow.Cells.Add(description);
                        continue;
                    }

                    var index = headerDataRowList.IndexOf(primaryTableSeriesDefinition.HeaderId);
                    var value = headerTable.Data.Columns[headerTableSeriesDefinition.Id]
                        .GetValueHelper(headerTableSeriesDefinition.Type, index);
                    headerRow.Cells.Add(value.ToString());
                }

                headerRow.isHeader = true;
                table.Rows.Add(headerRow);
            }

            // Constructs the column data
            for (int i = 0; i < rowCount; i++)
            {
                var dataRow = new Row {Cells = new List<string>()};
                foreach (var columnId in columnIds)
                {
                    dataRow.Cells.Add(Convert.ToString(primaryTable.Data.Columns[columnId]
                        .GetValueHelper(primaryTable.Definition.Columns.First(c => c.Id == columnId).Type, i)));
                }

                table.Rows.Add(dataRow);
            }

            var metadataItems = primaryTable.Data.Metadata.Items;
            var tableMetadataLocations = primaryTable.Data.Metadata.Locations.Table;

            foreach (var location in tableMetadataLocations)
            {
                table.Metadata.Add(metadataItems[location].Name, metadataItems[location].StringValue);
            }

            return table;
        }
    }
}

File: python/src/fds/protobuf/stach/extensions/v1/ColumnOrganizedStachExtension.py
import pandas as pd
from fds.protobuf.stach.Package_pb2 import Package

from fds.protobuf.stach.extensions.IStachExtension import IStachExtension
from fds.protobuf.stach.extensions.v1.StachUtilities import StachUtilities


class ColumnOrganizedStachExtension(IStachExtension):
    """The purpose of this class is to provide the helper methods for converting Stach to Tabular format"""

    def __init__(self, package: Package):
        self.package = package

    def convert_to_dataframe(self):
        tables = list()
        for primary_table_id in self.package.primary_table_ids:
            tables.append(ColumnOrganizedStachExtension.__generate_table(self.package, primary_table_id))
        return tables

    def get_metadata(self):
        metadata = list()
        for primary_table_id in self.package.primary_table_ids:
            metadata.append(self.__generate_metadata(self.package, primary_table_id))
        return metadata

    @staticmethod
    def __generate_table(package_response, primary_table_id):
        """
        The purpose of this function is to generate Tables for the provided stach data through the package.

        :param package_response: Stach Data which is represented as a Package object.
        :param primary_table_id: Refers to the id for a particular table inside a package.
        :return: Returns the generated Table from the package provided.
        """
        if isinstance(package_response, Package):
            primary_table = package_response.tables[primary_table_id]
            header_id = primary_table.definition.header_table_id
            header_table = package_response.tables[header_id]
            row_count = len(primary_table.data.rows)

            headers = list(list())
            # Constructs the column headers by considering dimension columns and header rows
            for series_definition_column in header_table.definition.columns:
                header_row = list()
                for primary_table_definition in primary_table.definition.columns:
                    if (primary_table_definition.is_dimension == True):
                        header_row.append(primary_table_definition.description or primary_table_definition.name)
                        continue
                    headerColumnId = primary_table_definition.header_id
                    indexOfHeaderColumnId = [index for index in range(len(header_table.data.rows)) if
                                             header_table.data.rows[index].id == headerColumnId][0]
                    val = str(StachUtilities.get_value_helper(header_table.data.columns[series_definition_column.id],
                                                              series_definition_column.type, indexOfHeaderColumnId,
                                                              series_definition_column.format.null_format))
                    header_row.append(val)
                headers.append(header_row)

            data = list(list())
            # Constructs the column data
            for i in range(0, row_count, 1):
                data_row = list()
                for series_definition_column in primary_table.definition.columns:
                    data_row.append(str(
                        StachUtilities.get_value_helper(primary_table.data.columns[series_definition_column.id],
                                                        series_definition_column.type, i,
                                                        series_definition_column.format.null_format)))
                data.append(data_row)

            if len(header_table.definition.columns) > 1:
                data_frame = pd.DataFrame(data=data)
                data_frame.columns = pd.MultiIndex.from_arrays(headers)
            else:
                data_frame = pd.DataFrame(data=data, columns=headers[0])

            return data_frame

        else:
            ValueError("Response data passed should be of package type.")

    @staticmethod
    def __generate_metadata(package_response, primary_table_id):
        """
        The purpose of this function is to generate the metadata for the provided stach data through the package.

        :param package_response: Stach Data which is represented as a Package object.
        :param primary_table_id: Refers to the id for a particular table inside a package.
        :return: Returns the generated metadata from the package provided.
        """
        primary_table = package_response.tables[primary_table_id]
        metadata = {}

        for location in primary_table.data.metadata.locations.table:
            name = primary_table.data.metadata.items[location].name

            stringval = primary_table.data.metadata.items[location].string_value
            values = stringval.split("|")
            
            metadata[name] = values
        return metadata
